src/Task.cc

```diff
@@ -2200,17 +2251,35 @@ static void set_up_process(Session& session, const ScopedFd& err_fd) {
   // which would be crazy ... though we could fix it by dynamically
   // assigning RR_RESERVED_ROOT_DIR_FD.)
   if (!running_under_rr()) {
+    NativeArch::rlimit old_limits, new_limits;
+    prlimit(0, RLIMIT_NOFILE, NULL, (rlimit*)&old_limits);
+
+    if (old_limits.rlim_cur < 1024 || old_limits.rlim_max < 1024) {
+      spawned_child_fatal_error(err_fd, "rr requires at least 1024 fds");
+    }
+
+    new_limits.rlim_max = old_limits.rlim_max;
+    new_limits.rlim_cur = old_limits.rlim_max;
+    prlimit(0, RLIMIT_NOFILE, (rlimit*)&new_limits, NULL);
+
+    int desired_fd = pick_reserved_fd_boundary(new_limits.rlim_cur);
+    if (0 > desired_fd) {
+      spawned_child_fatal_error(err_fd, "error picking fd boundary");
+    }
+
     /* CLOEXEC so that the original fd here will be closed by the exec that's
      * about to happen.
      */
     int fd = open("/", O_PATH | O_DIRECTORY | O_CLOEXEC);
     if (0 > fd) {
       spawned_child_fatal_error(err_fd, "error opening root directory");
     }
-    if (RR_RESERVED_ROOT_DIR_FD != dup2(fd, RR_RESERVED_ROOT_DIR_FD)) {
-      spawned_child_fatal_error(err_fd,
-                                "error duping to RR_RESERVED_ROOT_DIR_FD");
+    if (desired_fd != dup2(fd, desired_fd)) {
+      spawned_child_fatal_error(err_fd, "error duping to desired_fd");
     }
+
+    old_limits.rlim_cur = std::min<NativeArch::rlim_t>(old_limits.rlim_cur, desired_fd);
+    prlimit(0, RLIMIT_NOFILE, (rlimit*)&old_limits, NULL);
```

Comment:
> That would break anything that expected to be able to close all its own fds this way.

Good point.

> The tracee can see magical holes in the fd space that are unused but it can't dup to/etc.

True, though that's an existing issue that hasn't caused any problems so far. We could avoid it --- well, reduce it --- by allocating rr fds right at the top of the available range.